p2p/proto/protocols.md

```diff
@@ -1,11 +1,259 @@
+
+It is important not to validate the commitment of receipts until this is fixed because then the node
+can get stuck if the peer that gave the header is malicious and the peer that gives receipts is honest.
+
+### State Diff
+The state diff protocol is used to download the state diffs created by a range of blocks.
+It can be used to avoid running the transactions for computing the state diff.
+
+Its name for negotiation is `/starknet/state_diffs/1`
+
+In order to verify the state diff, the node needs to calculate its hash and compare it to the
+`state_diff_commitment` field in the block header. The structure of the hash is detailed [here](https://community.starknet.io/t/introducing-p2p-authentication-and-mismatch-resolution-in-v0-12-2/97993)
+
+#### Message structure
+Each single message is either a fin, a [ContractDiff](./state.proto) or a [DeclaredClass](./state.proto).
+
+A ContractDiff represents all the changes made for a given contract. This includes:
+* Storage diffs - Represented by the `values` field.
+* Nonce updates - Represented by the `nonce` field if present.
+* Deployed contracts - Represented by the `class_hash` field if present. If it's present it means
+that this contract was deployed or replaced in this block.
+
+A DeclaredClass represents a declared class. If the class is Cairo1 then the `compiled_class_hash`
+field will be present
+
+There's no guarantee on what ordering will the messages have. The only constraints are:
+1. If the same contract has multiple `ContractDiff`s, it's because they don't fit the
+[limit](#limits) for a single message.
+2. If that happens, then all `ContractDiff`s for the same contract need to appear continuously.
+
+#### Length
+As we've seen in other protocols like [Transactions](#transactions), The node needs to know the
+amount of data to expect in order to reject malicious peers trying to DDoS the node.
+
+In order to do that, the node needs to look at the block header at the field `state_diff_length`.
+This field is equal to
+`num_storage_diffs + num_nonce_updates + num_deployed_contracts + num_declared_classes`
+
+Then, whenever the node receives a [StateDiffResponse](./state.proto), it needs to add a number to
+a counter and declare the peer malicious if the counter becomes bigger than `state_diff_length`:
+* For DeclaredClass, the counter should increase by 1.
+* For ContractDiff, the counter should increase by the length of the field `values`,
+and by 1 if the field `class_hash` is present, and by 1 if the field `nonce` is present.
+
+##### Important Note
+Right now, the `state_diff_length` is not part of the block hash. Until this is fixed, it's value is
+trusted.
+
+In order for the node to not be stuck if the peer that gave it the header is malicious, the node
+should not validate the amount of state diff parts it got for a block and trust that the inbound
+peer won't try to DDoS it (and rely on the [limits](#limits) for DDoS prevention).
```

Comment:
not clear why here the node should not validate and in other places it should. 

Also, i'm not very comfortable having guidelines on how a node should behave with its peers. We should tell nodes how to validate the data. Whether they treat the peer as malicious or not is their choice.